### PR TITLE
Combined Reid and Tracking Model & Some Utils

### PR DESCRIPTION
**Pull request by darkushin**
- Added a new model under `models` which uses mmtracking and replaces the ids given by mmtrack with the ids computed using fast-reid.
- Added the build_dataset.py script to `utils`. This script allows the creation of a new dataset in a convenient way.
- Added a new utils.py file to the `DataProcessing` folder. This script should hold functions that are needed to manipulate the data, such as renaming images in a given path, etc.

### CODE REVIEW

File: DataProcessing/utils.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    """
+    Convert all images in the given path from its current malformed format to the `xxxx_c1_f1234567.jpg` format which is
+     the correct format for the DukeMTMC dataset.
+     Change this function according to the current corrections you need to do.
+    """
+    for im in os.listdir(os.path.join(path)):
+        if '.jpg' not in im:
+            continue
+        # new_im_name = im.split('.jpg')[0]
+        new_im_name = im.replace('c1', 'c2')
```
Comment (bar371) [inline DataProcessing/utils.py]: perhaps create a constant QUERY_CAMERA with value > 2 for the replace?

Comment (bar371) [inline DataProcessing/utils.py]: and should this recieve cam id when we would like to support 2 camera input?

File: models/model_runner.py [hunk 2]
```diff
@@ -86,8 +94,11 @@ def runner():
                                                 acc_threshold=args.acc_treshold,
                                                 capture_index=args.capture_index)
 
-    elif 're-id' in args.action:
+    elif args.action in REID_ACTIONS:
```
Comment (bar371) [inline models/model_runner.py:110]: Nice consistency!

File: models/track_and_reid_model.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    for (feat, pid, camid) in tqdm.tqdm(reid_model.run_on_loader(data), total=len(data)):
+        feats.append(feat)
+        pids.extend(pid)
+        camids.extend(camid)
+
+    print("The size of test gallery is", len(pids))
+    feats = torch.cat(feats, dim=0)
+    g_feat = feats
+    g_pids = np.asarray(pids)
+    # g_camids = np.asarray(camids)  # todo: remove?
```
Comment (bar371) [inline models/track_and_reid_model.py]: I think we should not remove it. We will use it once we start working with both cameras

File: models/track_and_reid_model.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    return feats, g_feat, g_pids
+
+
+def find_best_match(q_feat, g_feat, g_pids):
+    """
+    Given feature vectors of the query images, return the ids of the images that are most similar in the test gallery
+    """
+    features = F.normalize(q_feat, p=2, dim=1)
```
Comment (bar371) [inline models/track_and_reid_model.py:70]: I recall q_feat and g_feat are coming normalized already from the original re-id model right? if so we're good.

File: models/track_and_reid_model.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    distmat = 1 - torch.mm(features, others.t())
+
+    distmat = distmat.numpy()
+    best_match_in_gallery = np.argmin(distmat, axis=1)
+    return g_pids[best_match_in_gallery]
+
+
+def tracking_inference(tracking_model, img, frame_id, acc_threshold=0.8):
+    result = inference_mot(tracking_model, img, frame_id=frame_id)
+    # result['track_bboxes'] = result.get('track_results')
```
Comment (bar371) [inline models/track_and_reid_model.py]: if its working without the hack-fix lets remove it

File: models/track_and_reid_model.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    acc = result['track_results'][0][:, -1]
+    mask = np.where(acc > acc_threshold)
```
Comment (bar371) [inline models/track_and_reid_model.py:82]: acc_threshold - do we send it as a hyper param for tunning in the original model_runner? if not add it

File: models/track_and_reid_model.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    result['track_results'][0] = result['track_results'][0][mask]
+    result['bbox_results'][0] = result['bbox_results'][0][mask]
+    return result
+
+
+def reid_inference(reid_model, img, result, frame_id, crops_folder=None):
+    crops_bboxes = result['track_results'][0][:, 1:-1]
+    crops_imgs = mmcv.image.imcrop(img, crops_bboxes, scale=1.0, pad_fill=None)
+    q_feat = torch.empty((len(crops_imgs), 2048))
```
Comment (bar371) [inline models/track_and_reid_model.py:91]: note the 2048 here, this is exactly the size of the feature vector. we should have something a bit more stable if we will like to change the architecture of the fast-re-id. I would put crop_images[0].shape[1] (should give the size of the feature vector)

File: models/track_and_reid_model.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+    print("The size of test gallery is", len(pids))
+    feats = torch.cat(feats, dim=0)
+    g_feat = feats
+    g_pids = np.asarray(pids)
+    # g_camids = np.asarray(camids)  # todo: remove?
+    return feats, g_feat, g_pids
+
+
+def find_best_match(q_feat, g_feat, g_pids):
```
Comment (bar371) [inline models/track_and_reid_model.py]: For now this is fine, but In the future when we'll want to combine multi model we'll either have a diff class for it, or make into find_best_match_according_to_re_id model

**Review by bar371 — CHANGES_REQUESTED** [review]

File: DataProcessing/utils.py (hunk 1 above)
Comment (darkushin) [reply DataProcessing/utils.py]: For both comments:
This utils file is supposed to help you when you want to do a very specific action on ALREADY EXISTING data. I mean, that given some data folder you want to change (for example, I saw a folder from cam 2 we mistakenly labeled with c1) you will have functions in this file that will help you to quickly fix it, but you will have to manually set the changes you need. Hence, I think there is no need for constants here as they aren't really constant.

Comment (darkushin) [reply DataProcessing/utils.py]: see above comment

File: models/track_and_reid_model.py (hunk 3 above)
Comment (darkushin) [reply models/track_and_reid_model.py]: you right, I will leave it there

File: models/track_and_reid_model.py (hunk 4 above)
Comment (darkushin) [reply models/track_and_reid_model.py:70]: No, this was also the bug we fixed in the visualization - they didn't come normalized

File: models/track_and_reid_model.py (hunk 5 above)
Comment (darkushin) [reply models/track_and_reid_model.py]: Will remove, hope it will work for you the same

File: models/track_and_reid_model.py (hunk 7 above)
Comment (darkushin) [reply models/track_and_reid_model.py:91]: As you said it is the size of the feature vector, `crop_images` holds the different crops we extract from the image and its shape is the shape of every crop and not of the feature vector. There is no easy way to know here what the size of the feature vector is, but I don't think it is that terrible - once we add new features, this will be the only place we need to change the dimensions

**Review by bar371 — APPROVED** [review]
